### PR TITLE
fix: ensure `credits_available` API returns proper policies

### PR DESCRIPTION
**Pull request by adamstankiewicz**
[ENT-7708](https://2u-internal.atlassian.net/browse/ENT-7708)

Unblocks [ENT-7741](https://2u-internal.atlassian.net/browse/ENT-7741).

### CODE REVIEW

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py [hunk 1]
```diff
@@ -343,8 +343,7 @@ def policies_with_credit_available(self, enterprise_customer_uuid, lms_user_id):
         Return policies with credit availble, associated with the given customer, and redeemable by the given learner.
         """
         policies = []
-        # TODO: maybe we should use self.get_queryset()?
-        all_policies_for_enterprise = SubsidyAccessPolicy.objects.filter(
+        all_policies_for_enterprise = self.get_queryset().filter(
```
Comment (iloveagent57) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:346]: hah, real answer is always in the comments :)

File: enterprise_access/apps/subsidy_access_policy/models.py [hunk 2]
```diff
@@ -803,7 +879,18 @@ def can_redeem(self, lms_user_id, content_key, skip_customer_user_check=False):
...
+        Determine whether a learner has credit available for the subsidy access policy.
+        """
+        is_credit_available = super().credit_available(lms_user_id, skip_customer_user_check)
+        if not is_credit_available:
+            return False
+
+        if self.per_learner_spend_limit is None:
+            return True
+
+        # Validate whether learner has enough remaining balance (spend) for this policy.
```
Comment (iloveagent57) [inline enterprise_access/apps/subsidy_access_policy/models.py:893]: I think `self.remaining_balance_per_user()` needs to account for negative aggregates before comparing to the per-learner spend limit.

**Review by iloveagent57 — COMMENTED** [review]
Great changes, I think there's one more place to account for negative transaction quantities.

File: enterprise_access/apps/subsidy_access_policy/models.py [hunk 3]
```diff
@@ -486,6 +485,72 @@ def can_redeem(self, lms_user_id, content_key, skip_customer_user_check=False):
...
+
+        # verify associated subsidy is current (non-expired)
+        try:
+            if not self.is_subsidy_active:
+                logger.info('[credit_available] SubsidyAccessPolicy.subsidy_record() returned inactive subsidy')
+                return False
+        except requests.exceptions.HTTPError as exc:
+            # when associated subsidy is soft-deleted, the subsidy retrieve API raises an exception.
+            logger.info('[credit_available] SubsidyAccessPolicy.subsidy_record() raised HTTPError: %s', exc)
+            return False
```
[the remark this answers is not in the document]
Comment (adamstankiewicz) [reply enterprise_access/apps/subsidy_access_policy/models.py:540]: Yeah, I agree it'd be a nice addition! It also relates to the other idea we've discussed a few times in the past of de-activating policies where the associated subsidy has expired.

File: enterprise_access/apps/subsidy_access_policy/models.py (hunk 2 above)
Comment (adamstankiewicz) [reply enterprise_access/apps/subsidy_access_policy/models.py:893]: @iloveagent57 Good catch! I've updated to account for negative aggregates here and the associated tests in the most recent commit.

**Review by iloveagent57 — APPROVED** [review]

**Pull request merged**